### PR TITLE
feat: move supported features to the application service

### PR DESCRIPTION
**Pull request by SimonRichardson**
The supported features are charm-based feature set, so they should live in the application service. We can just hang this off from the service until we have full end-to-end paths for models and charms.

We can start to drop the cloud and credential services in the application domain, as we're slowly dropping the need to invoke a provider directly.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

```sh
$ juju bootstrap lxd test
$ juju add-model default
$ juju deploy nvidia-gpu-operator --channel 1.29/stable
ERROR Charm feature requirements cannot be met:
  - charm requires feature "k8s-api" but model does not support it

Feature descriptions:
  - "k8s-api": the Kubernetes API lets charms query and manipulate the state of API objects in a Kubernetes cluster

For additional information please see: https://juju.is/docs/olm/supported-features
ERROR failed to deploy charm "nvidia-gpu-operator"
```

## Links

**Jira card:** [JUJU-6624](https://warthogs.atlassian.net/browse/JUJU-6624)

[JUJU-6624]: https://warthogs.atlassian.net/browse/JUJU-6624?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: domain/application/service/application.go [hunk 1]
```diff
@@ -782,10 +786,81 @@ func (s *ApplicationService) GetApplicationScalingState(ctx context.Context, app
...
+type Provider interface {
+	environs.SupportedFeatureEnumerator
+}
+
+// ProviderApplicationService defines a service for interacting with the underlying
+// model state.
+type ProviderApplicationService struct {
+	ApplicationService
+
+	modelID            coremodel.UUID
```
Comment (tlm) [inline domain/application/service/application.go:809]: Are we passing this in via the service factory or pulling it off the DB every time? Confused which method we should be taking now?

Comment (SimonRichardson) [reply domain/application/service/application.go:809]: Model id is service factory.

**Review by hmlanigan — APPROVED** [review]
LGTM, qa went well, mostly questions and nits on comments.

**Pull request merged**